File: DotNetGotchas/CSharp/ParamThreadSafety/Modified3/ThreadPassingParams/ALib/SomeClass.cs
//SomeClass.cs part of ALib.dll
using System;
using System.Threading;

namespace ALib
{
	public class SomeClass
	{
		private void Method1(int val)
		{
			// Some operation takes place here
			Console.WriteLine(
				"Method1 runs on Thread {0} with {1}",
				AppDomain.GetCurrentThreadId(), val);
		}

		//...
		class CallMethod1Helper
		{
			private SomeClass theTarget;
			private int theValue;
			public CallMethod1Helper(int val, SomeClass target)
			{
				theValue = val;
				theTarget = target;
			}

			private void CallMethod1()
			{
				theTarget.Method1(theValue);
			}

			public void Run()
			{
				new Thread(
					new ThreadStart(CallMethod1)).Start();
			}
		}

		public void DoSomething(int val)
		{
			// Some operation...

			// Want to call Method1 in different thread
			// from here?
			CallMethod1Helper helper = new CallMethod1Helper(
				val, this);
			helper.Run();

			// Some operation...
		}
	}
}

File: DotNetGotchas/VBDN/ParamThreadSafety/Modified3/ThreadPassingParams/ALib/SomeClass.vb
'SomeClass.vb part of ALib.dll
Imports System.Threading

Public Class SomeClass

    Private Sub Method1(ByVal val As Integer)
        Console.WriteLine( _
            "Method1 runs on Thread {0} with {1}", _
            AppDomain.GetCurrentThreadId(), val)

        ' Some operation takes place here
    End Sub

    '...
    Class CallMethod1Helper
        Private theTarget As SomeClass
        Private theValue As Integer

        Public Sub New(ByVal val As Integer, ByVal target As SomeClass)
            theValue = val
            theTarget = target
        End Sub

        Private Sub CallMethod1()
            theTarget.Method1(theValue)
        End Sub

        Public Sub Run()
            Dim theThread As New Thread(AddressOf CallMethod1)
            theThread.Start()
        End Sub
    End Class


    Public Sub DoSomething(ByVal val As Integer)
        ' Some operation...

        ' Want to call Method1 in different thread
        ' from here?
        Dim helper As New CallMethod1Helper(val, Me)
        helper.Run()

        ' Some operation...
    End Sub
End Class
